### ISSUE TITLE
Unable to upload file more than 300MB

### ISSUE THREAD

**Issue by supportreq** [nextcloud/android#329]
### Actual behaviour
- I am trying to upload a 1.2GB file from my Android Client
- The Client uploads it till around 100MB/ 150MB and breaks asks for a upload retry
- One file was able to upload around 250MB and than it failed
- Uploading the same file from the browser on android phone works fine
### Expected behaviour
- Upload should complete with out any issue from the android app
### Steps to reproduce
1. 
2. 
3. 
### Environment data

Android version: Android 6.0.1

Device model: OnePlus One

Stock or customized system: CM Nightly

Nextcloud app version: 1.3.0.. I tried with OwnCloud Android app as well

Nextcloud server version: Actually using OwnCloud 8.2.2

**Comment by AndyScherzinger** [comment]
Hi @supportreq thanks for reporting this. Are you uploading on a wifi or cellular connection and what happens after a retry? Does it (at some point) complete the upload or does it fail straight away after it failed once?

**Comment by supportreq** [comment]
@AndyScherzinger I am uploading using Wifi.. The upload restarts from 0% and again fails after sometime..

**Comment by AndyScherzinger** [comment]
@supportreq Do you know on which Android version CM nightly is based?

**Comment by supportreq** [comment]
Android marshmallow.. 6.0.1

**Comment by iskradelta** [comment]
Seen this problem with less < 10MB files, when uploading a thousand or so camera photos. Some just wouldnt upload no matter how many retries.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky any idea? In general I think we should add the log as a general feature which is deactivated by default and can be activated by the user having it hidden deep, deep down in the settings screen. What do you think @jancborchardt ? I now we didn't include it up to now since it doesn't feel  cosumer product-ish but the number of non-reproducable issues is increasing and the log information might be crucial.

**Comment by iskradelta** [comment]
Got it! At least 1 bug reproducable now, but not in code as a test-case yet.

User sees "Unkown error" "Upload failed", for the same file, no matter how many retries, only for some files.

What is happening is" upon retries of sending chunks when other chunks have been successful", the last chunk always got the content-length as CHUNK_SIZE 1024000, because it was calculating from channel.position before the position was moved, the solution is to calculate content-length using the offset. It works for all other files because the channel position is updated as we send the file The reproducable error only occurs when chunked upload and for some reason the first attempt at uploading the whole file fails. A little more in-depth theory below:

See this code is here in ChunkedUploadRemoteFileOperation.java (android-library), it reads in successfulChunks from savedPreferences (where we save chunks successfully sent before), then if we see chunk 1 or 2 has been sent, the for loop continues, and in doing so it sets offset += CHUNK_SIZE, later setOffset is called and the http put method executed which eventually calles writeBytes where the channel is set to position of offset to read from, but before this happened, httplib called getContentLength (return Math.min(chunksize, channel.size - channel.position) were we used the channel position (which was at 0 since this is resumption operation), and so the CHUNK_SIZE was smaller than file-size and thats returned, even though we have less content to serve! So anyhow, the request fires off, the web-server waits for more data to come but it never does. So it times out, we get an IOException somewhere and decode it to "Unrecovered transport exception" and tell the user "Unknown error".

So the solution is return Math.min(chunsize, channel.size - mOffset).

Tried this fix, and it succeeds! 

Now then, to figure out why the hell it failed on the first attempt of uploading - while other files succeed this one didnt - I dont know thats another bug, or just random drop of connectivity/wifi/whatver/spooky xrays.
